xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -83,15 +98,39 @@
   @VisibleForTesting
   static boolean enableTimeout =
       Boolean.parseBoolean(System.getenv("GRPC_XDS_EXPERIMENTAL_ENABLE_TIMEOUT"));
+  @VisibleForTesting
+  static boolean enableFaultInjection =
+      Boolean.parseBoolean(System.getenv("GRPC_XDS_EXPERIMENTAL_FAULT_INJECTION"));
+  @VisibleForTesting
+  static final Metadata.Key<String> DOWNSTREAM_NODE_KEY =
+      Metadata.Key.of("x-envoy-downstream-service-node", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_HTTP_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_GRPC_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-grpc-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  static final AtomicLong activeFaultInjectedStreamCounter = new AtomicLong();
 
   private final XdsLogger logger;
   private final String authority;
   private final ServiceConfigParser serviceConfigParser;
   private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService scheduler;
   private final XdsClientPoolFactory xdsClientPoolFactory;
   private final ThreadSafeRandom random;
   private final ConcurrentMap<String, AtomicInteger> clusterRefs = new ConcurrentHashMap<>();
   private final ConfigSelector configSelector = new ConfigSelector();
+  private final AtomicLong activeFaultInjectedStreams;
```

Comment:
This is very weird (as well as the DI in provider). 

My suggestion is still to put this atomic in `SharedCallCounterMap` and inject an `SharedCallCounterMap` instance. So that we turn globals to dependencies.

If you still want to just have the static atomic declared here, you could make it non-final and have a setter to override it for testing. Anyway, the pattern of having a non-static atomic mirroring the static one is odd.

Reply:
I don't think messing up `SharedCallCounterMap` is a good idea. I'm okay to taken the non-final filed pattern although it's effectively the same. Whatever.

Reply:
See my comment at https://github.com/grpc/grpc-java/pull/7870#discussion_r578046944